pkg/cmd/server/origin/handlers.go

```diff
@@ -291,3 +296,92 @@ func assetServerRedirect(handler http.Handler, assetPublicURL string) http.Handl
 		handler.ServeHTTP(w, req)
 	})
 }
+
+func (c *MasterConfig) impersonationFilter(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		requestedSubject := req.Header.Get("Impersonate-User")
+		if len(requestedSubject) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		resource, namespace, name, err := parseRequestedSubject(requestedSubject)
+		if err != nil {
+			forbidden(err.Error(), nil, w, req)
+			return
+		}
+
+		ctx, exists := c.RequestContextMapper.Get(req)
+		if !exists {
+			forbidden("context not found", nil, w, req)
+			return
+		}
+
+		actingAsAttributes := &authorizer.DefaultAuthorizationAttributes{
+			Verb:         "impersonate",
+			APIGroup:     resource.Group,
+			Resource:     resource.Resource,
+			ResourceName: name,
+		}
+		authCheckCtx := kapi.WithNamespace(ctx, namespace)
+
+		allowed, reason, err := c.Authorizer.Authorize(authCheckCtx, actingAsAttributes)
+		if err != nil {
+			forbidden(err.Error(), actingAsAttributes, w, req)
+			return
+		}
+		if !allowed {
+			forbidden(reason, actingAsAttributes, w, req)
+			return
+		}
+
+		switch resource {
+		case kapi.Resource("serviceaccounts"):
+			newUser := &user.DefaultInfo{
+				Name:   serviceaccount.MakeUsername(namespace, name),
+				Groups: serviceaccount.MakeGroupNames(namespace, name),
+			}
+			newUser.Groups = append(newUser.Groups, "system:authenticated")
+			c.RequestContextMapper.Update(req, kapi.WithUser(ctx, newUser))
+
+		case userapi.Resource("users"):
+			newUser := &user.DefaultInfo{
+				Name: name,
+			}
+			groups, err := c.GroupCache.GroupsFor(name)
+			if err == nil {
+				for _, group := range groups {
+					newUser.Groups = append(newUser.Groups, group.Name)
+				}
+			}
+			newUser.Groups = append(newUser.Groups, "system:authenticated", "system:authenticated:oauth")
+			c.RequestContextMapper.Update(req, kapi.WithUser(ctx, newUser))
+
+		default:
+			forbidden(fmt.Sprintf("%v is an unhandled resource for acting-as", resource), nil, w, req)
+			return
+		}
+
+		newCtx, _ := c.RequestContextMapper.Get(req)
+		oldUser, _ := kapi.UserFrom(ctx)
+		newUser, _ := kapi.UserFrom(newCtx)
+		httplog.LogOf(req, w).Addf("%v is acting as %v", oldUser, newUser)
+
+		handler.ServeHTTP(w, req)
+	})
+}
+
+func parseRequestedSubject(requestedSubject string) (unversioned.GroupResource, string, string, error) {
+	subjects := authorizationapi.BuildSubjects([]string{requestedSubject}, nil, func(s string, b bool) (bool, string) { return true, "" }, func(s string, b bool) (bool, string) { return true, "" })
```

Comment:
a comment on what these functions are doing would be helpful... this is a little impenetrable